Add Utilities::MPI::gather

BenBrands opened the pull request.

davydden commented on include/deal.II/base/mpi.templates.h:
```diff
@@ -324,6 +324,71 @@ namespace Utilities
+      // Exchanging the size of each buffer
+      MPI_Gather(&n_local_data, 1, MPI_INT,
+                 size_all_data.data(), 1, MPI_INT,
+                 0,comm);
```
maybe instead of hard-coding `0`, add 
```
const unsigned int root = 0;
```
at the top and use it throughout.

masterleinad replied:
Actually, I would prefer if the id of the receiving MPI process is a parameter (with default value 0).

jppelteret commented on include/deal.II/base/mpi.templates.h:
```diff
@@ -324,6 +324,71 @@ namespace Utilities
+      const auto n_procs = dealii::Utilities::MPI::n_mpi_processes(comm);
+      const auto my_rank = dealii::Utilities::MPI::this_mpi_process(comm);
+
+      std::vector<char> buffer = Utilities::pack(object);
```
Could this be made `const`?

masterleinad commented:
The only place where we currently use `MPI_Gather` is in `source/base/index_set.cc:645`. Would you mind updating that place with this new function?

davydden commented on include/deal.II/base/mpi.templates.h:
```diff
@@ -324,6 +324,71 @@ namespace Utilities
+      if (my_rank==0)
+        received_unrolled_buffer.resize(rdispls.back() + size_all_data.back());
+
+      MPI_Gatherv(buffer.data(), n_local_data, MPI_CHAR,
```
please add
```
const int ierr = ...
AssertThrowMPI(ierr);
```

BenBrands commented:
Added tests for `Utilities::MPI::gather`.
@masterleinad: As you asked for, I replaced `MPI_gather` by `Utilities::MPI::gather`. Therefore, I had to include [include/deal.II/base/mpi.templates.h](https://github.com/dealii/dealii/blob/master/include/deal.II/base/mpi.templates.h)!

davydden commented:
./run-tests

bangerth commented on include/deal.II/base/mpi.h:
```diff
@@ -534,6 +534,27 @@ namespace Utilities
+     * argument.
+     *
+     * @param[in] comm MPI communicator.
+     * @param[in] object_to_send An object to send to the root process
```
can you add a description also of the defaulted last argument?

bangerth commented on include/deal.II/base/mpi.templates.h:
```diff
@@ -324,6 +324,77 @@ namespace Utilities
+    template<typename T>
+    std::vector<T> gather(const MPI_Comm &comm,
+                          const T &object,
+                          const unsigned int root_process)
```
I think this function should go into the `.h` file. We can't predict the type `T` for any reasonable subset of cases.

BenBrands replied:
Correct me if I'm wrong, but I thought just the function declaration is in the `.h` file and the definition is placed in the `.templates.h` file.
That is at least the case for the other wrappers in [include/deal.II/base/mpi.h](https://github.com/dealii/dealii/pull/5784/files/d9d834765800fa76bc3b95b0cd5cac4ff45bcdb1#diff-51e42c602d7e746a09658ab4fae843cc), where you can also find `Utilities::MPI::all_gather`.

bangerth replied:
I think this falls under the last category of the discussion here: doc/doxygen/headers/instantiations.h

I tend to think that the `all_gather()` function has ended up in the `.templates.h` file by mistake. I'd accept a (separate) patch that moves it into the `.h` file, if you felt enterprising :-)

BenBrands replied:
@bangerth - I will have look where the other misplaced MPI functions are used.

bangerth commented on include/deal.II/base/mpi.h:
```diff
@@ -534,6 +534,27 @@ namespace Utilities
+     * @return A vector of objects, with size equal to the number of
+     *  processes in the MPI communicator. Each entry contains the object
+     *  received from the processor with the corresponding rank within the
+     *  communicator.
```
Can you document that this non-empty vector is really only returned on the *root* process, whereas on all other processors, it is an empty vector?

bangerth commented on source/base/index_set.cc:
```diff
@@ -637,15 +638,7 @@ IndexSet::is_ascending_and_one_to_one (const MPI_Comm &communicator) const
-                        &(global_dofs[0]), 1, DEAL_II_DOF_INDEX_MPI_TYPE, 0,
-                        communicator);
-  AssertThrowMPI(ierr);
+  std::vector<types::global_dof_index> global_dofs = Utilities::MPI::gather(communicator,first_local_dof,0);
```
I think you can make it `const`

masterleinad commented on include/deal.II/base/mpi.h:
```diff
@@ -573,6 +595,76 @@ namespace Utilities
+      const auto n_procs = dealii::Utilities::MPI::n_mpi_processes(comm);
+      const auto my_rank = dealii::Utilities::MPI::this_mpi_process(comm);
+
+      Assert(root_process >= 0 && root_process < n_procs,
```
Since `root_process` is a `unsigned int`, it is always greater or equal than 0. Would you remove that check?

masterleinad commented on include/deal.II/base/mpi.h:
```diff
@@ -573,6 +595,76 @@ namespace Utilities
+    {
+#ifndef DEAL_II_WITH_MPI
+      (void)comm;
+      std::vector<T> v(1, object_to_send);
```
In a build without `MPI`, `root_process` is unused. Would you mark it as "used" using
```
(void) root_process
```
?
